File: scripts/check_gates.py
"""CI gate checker for Python benchmark results."""

from __future__ import annotations

import argparse
import json
from pathlib import Path


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate benchmark gates.")
    parser.add_argument("--input", type=Path, required=True, help="Path to benchmark JSON")
    parser.add_argument("--min-accuracy", type=float, default=99.9)
    parser.add_argument("--max-p95-ms", type=float, default=1.0)
    args = parser.parse_args()

    data = json.loads(args.input.read_text(encoding="utf-8"))
    accuracy = data["accuracy"]
    spec = data["spec_sample"]

    failures: list[str] = []
    if float(accuracy["exact_case_accuracy_pct"]) < args.min_accuracy:
        failures.append(
            f"exact-case accuracy {accuracy['exact_case_accuracy_pct']:.6f}% < {args.min_accuracy:.3f}%"
        )
    if float(accuracy["per_tag_accuracy_pct"]) < args.min_accuracy:
        failures.append(f"per-tag accuracy {accuracy['per_tag_accuracy_pct']:.6f}% < {args.min_accuracy:.3f}%")
    if float(spec["regex_parse"]["p95_ms"]) >= args.max_p95_ms:
        failures.append(f"regex parse p95 {spec['regex_parse']['p95_ms']:.6f} ms >= {args.max_p95_ms:.6f} ms")
    if float(spec["regex_clean"]["p95_ms"]) >= args.max_p95_ms:
        failures.append(f"regex clean p95 {spec['regex_clean']['p95_ms']:.6f} ms >= {args.max_p95_ms:.6f} ms")

    if failures:
        print("CI gates: FAIL")
        for failure in failures:
            print(f"- {failure}")
        return 1

    print("CI gates: PASS")
    print(f"- exact-case accuracy: {accuracy['exact_case_accuracy_pct']:.6f}%")
    print(f"- per-tag accuracy: {accuracy['per_tag_accuracy_pct']:.6f}%")
    print(f"- regex parse p95: {spec['regex_parse']['p95_ms']:.6f} ms")
    print(f"- regex clean p95: {spec['regex_clean']['p95_ms']:.6f} ms")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/scripts/check_gates.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

function parseArg(flag: string, fallback: string): string {
  const idx = process.argv.indexOf(flag);
  if (idx >= 0 && process.argv[idx + 1]) {
    return process.argv[idx + 1];
  }
  return fallback;
}

function main(): number {
  const input = resolve(parseArg("--input", "results/benchmark_results.json"));
  const minAccuracy = Number.parseFloat(parseArg("--min-accuracy", "99.9"));
  const maxP95 = Number.parseFloat(parseArg("--max-p95-ms", "1.0"));

  const data = JSON.parse(readFileSync(input, "utf-8"));
  const accuracy = data.accuracy;
  const spec = data.spec_sample;

  const failures: string[] = [];
  if (accuracy.exact_case_accuracy_pct < minAccuracy) {
    failures.push(`exact-case accuracy ${accuracy.exact_case_accuracy_pct.toFixed(6)}% < ${minAccuracy.toFixed(3)}%`);
  }
  if (accuracy.per_tag_accuracy_pct < minAccuracy) {
    failures.push(`per-tag accuracy ${accuracy.per_tag_accuracy_pct.toFixed(6)}% < ${minAccuracy.toFixed(3)}%`);
  }
  if (spec.regex_parse.p95_ms >= maxP95) {
    failures.push(`regex parse p95 ${spec.regex_parse.p95_ms.toFixed(6)} ms >= ${maxP95.toFixed(6)} ms`);
  }
  if (spec.regex_clean.p95_ms >= maxP95) {
    failures.push(`regex clean p95 ${spec.regex_clean.p95_ms.toFixed(6)} ms >= ${maxP95.toFixed(6)} ms`);
  }

  if (failures.length > 0) {
    console.error("CI gates: FAIL");
    failures.forEach((failure) => console.error(`- ${failure}`));
    return 1;
  }

  console.log("CI gates: PASS");
  console.log(`- exact-case accuracy: ${accuracy.exact_case_accuracy_pct.toFixed(6)}%`);
  console.log(`- per-tag accuracy: ${accuracy.per_tag_accuracy_pct.toFixed(6)}%`);
  console.log(`- regex parse p95: ${spec.regex_parse.p95_ms.toFixed(6)} ms`);
  console.log(`- regex clean p95: ${spec.regex_clean.p95_ms.toFixed(6)} ms`);
  return 0;
}

process.exit(main());

